sympy/utilities/lambdify.py

```diff
@@ -905,6 +919,62 @@ def _module_present(modname, modlist):
             return True
     return False
 
+def _cse_homogeneous(exprs, **kwargs):
+    """
+    Same as ``cse`` but the reduced_exprs are expres in the same type as 'exprs'
+
+    Parameters
+    ==========
+
+    exprs : list of sympy expressions, or a single sympy expression
+        The expressions to reduce.
+    kwargs : The over argument of the ``cse`` function
+
+    Returns
+    =======
+
+    replacements : list of (Symbol, expression) pairs
+        All of the common subexpressions that were replaced. Subexpressions
+        earlier in this list might show up in subexpressions later in this
+        list.
+    reduced_exprs : list of sympy expressions
+        The reduced expressions with all of the replacements above.
+
+    Examples
+    ========
+
+    >>> from sympy.utilities.lambdify import _cse_homogeneous
+    >>> from sympy import Symbol, cos, Tuple
+    >>> x = Symbol("x")
+    >>> _cse_homogeneous("cos(x + 1)") # str
+    ([], 'cos(x + 1)')
+    >>> _cse_homogeneous(cos(x + 1)) # Basic
+    ([], cos(x + 1))
+    >>> _cse_homogeneous(1) # number
+    ([], 1)
+    >>> _cse_homogeneous(Tuple(1, x))
+    ([], (1, x))
+    >>> _cse_homogeneous([1, x])
+    ([], [1, x])
+    >>> _cse_homogeneous((1, x))
+    ([], (1, x))
+    >>> _cse_homogeneous({1, x})
+    ([], {1, x})
+    """
+    from sympy import cse, sympify, Matrix
+    from sympy.core.containers import Tuple
+
+    if isinstance(exprs, str):
+        replacements, (reduced_exprs,) = cse(sympify(exprs), **kwargs)
```

Comment:
What is the purpose of this helper function? The string parsing here can lead to issues. I would prefer to avoid it entirely. The remaining cases should be handled by cse() itself, if they aren't already.

Reply:
The next test failed:
```python
>>> lambdify("x, y, z", "z, x, y")(1, 2, 3) == (3, 1, 2)
```
It is maybe not the best mean to solve it, Do you prefer I make is fixing into the 'cse' function?